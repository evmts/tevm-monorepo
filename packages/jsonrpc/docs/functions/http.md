[**@tevm/jsonrpc**](../README.md) • **Docs**

***

[@tevm/jsonrpc](../globals.md) / http

# Function: http()

> **http**(`url`?, `config`?): `HttpTransport`

## Parameters

• **url?**: `string`

URL of the JSON-RPC API. Defaults to the chain's public RPC URL.

• **config?**: `HttpTransportConfig`

## Returns

`HttpTransport`

## Description

Creates a HTTP transport that connects to a JSON-RPC API.

## Source

node\_modules/.pnpm/viem@2.15.1\_bufferutil@4.0.8\_typescript@5.5.2\_utf-8-validate@6.0.4\_zod@3.23.8/node\_modules/viem/\_types/clients/transports/http.d.ts:44
