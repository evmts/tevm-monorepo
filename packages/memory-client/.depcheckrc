ignores: [
  "@ethereumjs/evm",
  "@ethereumjs/util",
  "@ethereumjs/block",
  "@ethereumjs/blockchain",
  "@ethereumjs/vm",
  "@tevm/actions-types",
  "@tevm/errors"
]
skip-missing: true
