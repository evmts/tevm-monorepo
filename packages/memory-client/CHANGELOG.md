# @tevm/client

## 1.0.0-next.23

### Minor Changes

- [#821](https://github.com/evmts/tevm-monorepo/pull/821) [`f2707baa92220f7848912037638ebad125dee539`](https://github.com/evmts/tevm-monorepo/commit/f2707baa92220f7848912037638ebad125dee539) Thanks [@0xNonCents](https://github.com/0xNonCents)! - Added Load State and Dump State to the API.

  These handlers allow one to read and write the entire tevm state similar to [load state and dump state in anvil](https://book.getfoundry.sh/reference/cli/anvil). This can be used to persist the state on disk or browser cache

### Patch Changes

- [#837](https://github.com/evmts/tevm-monorepo/pull/837) [`fd6d6aee21b8d72ab37d7b9117231f68812e2256`](https://github.com/evmts/tevm-monorepo/commit/fd6d6aee21b8d72ab37d7b9117231f68812e2256) Thanks [@roninjin10](https://github.com/roninjin10)! - Renamed createTevm createMemoryTevm

- [#837](https://github.com/evmts/tevm-monorepo/pull/837) [`fd6d6aee21b8d72ab37d7b9117231f68812e2256`](https://github.com/evmts/tevm-monorepo/commit/fd6d6aee21b8d72ab37d7b9117231f68812e2256) Thanks [@roninjin10](https://github.com/roninjin10)! - Changed the name of import('@tevm/memory-client').Tevm to MemoryTevm. This disambigouates it from the import('@tevm/actions-types').Tevm type that it extends.

- [#837](https://github.com/evmts/tevm-monorepo/pull/837) [`fd6d6aee21b8d72ab37d7b9117231f68812e2256`](https://github.com/evmts/tevm-monorepo/commit/fd6d6aee21b8d72ab37d7b9117231f68812e2256) Thanks [@roninjin10](https://github.com/roninjin10)! - Updated tevm to not proxy any json rpc requests it doesn't support. Proxying creates confusion because some methods operate off of the tevm state and others may be using the RPC state creating confusing mismatches. This means for now all unsupported rpc methods to a MemoryTevm will throw.

- [#862](https://github.com/evmts/tevm-monorepo/pull/862) [`f217fa4dc2f730cc109940ef36872229ae9f26d3`](https://github.com/evmts/tevm-monorepo/commit/f217fa4dc2f730cc109940ef36872229ae9f26d3) Thanks [@roninjin10](https://github.com/roninjin10)! - - Renamed MemoryTevm MemoryClient
  - Renamed TevmClient HttpClient
  - Replaced @tevm/actions-types package with @tevm/actions-types, @tevm/client-types, and @tevm/procedures-types packages
  - Moved errors to @tevm/errors
  - Moved bundler packages out of tevm and to @tevm/bundler package
  - Minimized packages exposed in tevm package
  - Fixed bug with missing types exports
- Updated dependencies [[`de81ac31460bb642dad401571ad3c1d81bdbef2d`](https://github.com/evmts/tevm-monorepo/commit/de81ac31460bb642dad401571ad3c1d81bdbef2d), [`f217fa4dc2f730cc109940ef36872229ae9f26d3`](https://github.com/evmts/tevm-monorepo/commit/f217fa4dc2f730cc109940ef36872229ae9f26d3), [`f2707baa92220f7848912037638ebad125dee539`](https://github.com/evmts/tevm-monorepo/commit/f2707baa92220f7848912037638ebad125dee539), [`37b936fd4a8095cd79875f5f1ca43f09442e653f`](https://github.com/evmts/tevm-monorepo/commit/37b936fd4a8095cd79875f5f1ca43f09442e653f)]:
  - @tevm/jsonrpc@1.0.0-next.23
  - @tevm/actions@1.0.0-next.23
  - @tevm/contract@1.0.0-next.23
  - @tevm/errors@1.0.0-next.23
  - @tevm/procedures@1.0.0-next.23

## 1.0.0-next.22

### Patch Changes

- Updated dependencies [[`8b3218b1`](https://github.com/evmts/tevm-monorepo/commit/8b3218b129ed43cf173a369cbe6b636365748e77), [`39a5b5e5`](https://github.com/evmts/tevm-monorepo/commit/39a5b5e52c704d1661b235b271e68129e7dc2a80), [`3b5f6729`](https://github.com/evmts/tevm-monorepo/commit/3b5f67291550b590dda16471059a05bd10fe324d), [`d514d111`](https://github.com/evmts/tevm-monorepo/commit/d514d111ff6b479fbbac07083477d59d70de1290), [`98d76506`](https://github.com/evmts/tevm-monorepo/commit/98d76506e5947678eb34127dcc6e4da7fa13cb68)]:
  - @tevm/vm@1.0.0-next.22
  - @tevm/viem@1.0.0-next.22
  - @tevm/contract@1.0.0-next.22
