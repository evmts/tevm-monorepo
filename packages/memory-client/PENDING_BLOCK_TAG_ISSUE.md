# Pending Block Tag Implementation Issue

## Problem

Tests related to the `blockTag: 'pending'` functionality fail with state root errors like:

```
Error: State root for 0xd2bcb7ee2771929cc89a69e60f23affd624526b9c69332f17b2ca48e9a1561a0 does not exist
```

## Root Cause

The issue appears to be in the implementation of `getPendingClient` function in `@tevm/actions/src/internal/getPendingClient.js`. The way deep copying works in the VM is causing state root references to break.

When the `getPendingClient` function creates a deep copy of the client with `client.deepCopy()` and then mines all transactions in the pool, some state root references are lost, leading to errors when trying to use the resultant client for operations.

## Current Approach

The current approach in implementing `blockTag: 'pending'` functionality is:

1. Deep copy the client to create a new one
2. Mine all transactions in the pool to create a simulated state
3. Execute the action on this simulated state with `blockTag: 'latest'`

This approach seems conceptually sound but has implementation issues related to state management.

## Failed Mitigation Attempts

1. Adding account initialization with balance before the tests that use `blockTag: 'pending'` - This didn't resolve the state root errors.

## Possible Solutions

1. **Temporary**: Skip tests requiring `blockTag: 'pending'` functionality to allow progress on other aspects of the codebase (current approach taken).

2. **Potential Fix**: Investigate the VM's deepCopy implementation, particularly how state roots are handled. The issue may be in one or more of:
   - `packages/vm/src/actions/deepCopy.js`
   - `packages/state/src/actions/deepCopy.js`
   - `packages/node/src/createTevmNode.js`

3. **Alternative Implementation**: Rather than using `deepCopy` + mine approach, consider creating a new implementation for `blockTag: 'pending'` that applies pending transactions directly to a temporary state or uses a different mechanism to simulate pending state.

## Next Steps

1. The pending block tag tests have been skipped for now to unblock other development work.
2. A deeper investigation into state management and deep copying in the VM is needed.
3. Consider refactoring the approach to handling pending block tags to avoid complex state copying mechanics.