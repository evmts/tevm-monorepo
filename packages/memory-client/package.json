{
	"name": "@tevm/memory-client",
	"version": "1.0.0-next.23",
	"private": false,
	"description": "Client implementations of Tevm",
	"keywords": [
		"solidity",
		"sol",
		"typescript",
		"web3",
		"blockchain"
	],
	"repository": {
		"type": "git",
		"url": "https://github.com/evmts/tevm-monorepo.git",
		"directory": "core/memory-client"
	},
	"license": "MIT",
	"contributors": [
		"Will Cory <willcory10@gmail.com>"
	],
	"type": "module",
	"exports": {
		"./package.json": "./package.json",
		".": {
			"import": {
				"types": "./types/index.d.ts",
				"default": "./dist/index.js"
			},
			"require": {
				"types": "./dist/index.d.cts",
				"default": "./dist/index.cjs"
			}
		}
	},
	"main": "dist/index.cjs",
	"module": "dist/index.js",
	"types": "types/index.d.ts",
	"files": [
		"dist",
		"src",
		"!src/**/*.spec.ts",
		"types"
	],
	"scripts": {
		"all": "pnpm i  && bun run build && bun lint && bun format && bun test:run && bun generate:docs && bun lint:deps && bun lint:package",
		"build": "nx run-many --targets=build:dist,build:types --projects=@tevm/memory-client",
		"build:dist": "tsup",
		"build:types": "tsup --dts-only && tsc --emitDeclarationOnly --declaration",
		"clean": "rm -rf node_modules && rm -rf artifacts && rm -rf dist && rm -rf cache",
		"format": "rome format . --write",
		"format:check": "rome format .",
		"generate:docs": "typedoc",
		"lint": "rome check . --apply-unsafe",
		"lint:check": "rome check . --verbose",
		"lint:deps": "bunx depcheck",
		"lint:package": "bunx publint --strict && attw --pack",
		"package:up": "pnpm up --latest",
		"test": "bun test --watch",
		"test:coverage": "bun test --coverage",
		"test:run": "bun test",
		"typecheck": "tsc --noEmit"
	},
	"dependencies": {
		"@ethereumjs/block": "^5.0.1",
		"@ethereumjs/blockchain": "^7.0.1",
		"@ethereumjs/common": "^4.1.0",
		"@ethereumjs/evm": "^2.1.0",
		"@ethereumjs/trie": "^6.0.1",
		"@ethereumjs/vm": "^7.1.0",
		"@tevm/actions": "workspace:^",
		"@tevm/contract": "workspace:^",
		"@tevm/errors": "workspace:^",
		"@tevm/jsonrpc": "workspace:^",
		"@tevm/procedures": "workspace:^",
		"viem": "^2.0.0"
	},
	"devDependencies": {
		"@ethereumjs/util": "^9.0.1",
		"@tevm/client-types": "workspace:^",
		"@tevm/predeploys": "workspace:^",
		"@tevm/procedures-types": "workspace:^",
		"@tevm/state": "workspace:^",
		"@tevm/tsconfig": "workspace:^",
		"@tevm/tsupconfig": "workspace:^",
		"abitype": "^1.0.0"
	},
	"publishConfig": {
		"access": "public"
	}
}
