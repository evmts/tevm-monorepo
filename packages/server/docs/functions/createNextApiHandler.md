[**@tevm/server**](../README.md) • **Docs**

***

[@tevm/server](../globals.md) / createNextApiHandler

# Function: createNextApiHandler()

> **createNextApiHandler**(`client`): `NextApiHandler`\<`any`\>

## Parameters

• **client**: [`Client`](../type-aliases/Client.md)

## Returns

`NextApiHandler`\<`any`\>

## Defined in

[packages/server/src/adapters/createNextApiHandler.js:16](https://github.com/evmts/tevm-monorepo/blob/main/packages/server/src/adapters/createNextApiHandler.js#L16)
