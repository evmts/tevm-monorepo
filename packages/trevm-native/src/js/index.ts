import { createLogger } from '@tevm/logger';
import { InternalError } from '@tevm/errors';
import type { StateManager } from '@tevm/state';
import type { Common } from '@tevm/common';
import type { Chain } from '@tevm/blockchain';

// Import native module
import { init_trevm, TrevmWrapper } from '../../dist/index.node';

export interface TrevmEvmOptions {
  common: Common;
  stateManager: StateManager;
  blockchain: Chain;
  customPrecompiles?: any[];
  profiler?: boolean;
  allowUnlimitedContractSize?: boolean;
  loggingLevel?: string;
}

export interface TrevmRunCallOptions {
  caller?: string;
  to?: string;
  data?: string;
  value?: bigint;
  gasLimit?: bigint;
  skipBalance?: boolean;
}

export interface TrevmRunCallResult {
  returnValue: string;
  executionGasUsed: string;
  error?: string;
}

export interface TrevmEvm {
  runCall(opts: TrevmRunCallOptions): Promise<TrevmRunCallResult>;
}

/**
 * Creates a trevm-based EVM instance
 * @param options EVM options
 * @returns A trevm EVM instance
 */
export const createTrevm = async ({
  common,
  stateManager,
  blockchain,
  customPrecompiles,
  profiler,
  allowUnlimitedContractSize,
  loggingLevel,
}: TrevmEvmOptions): Promise<TrevmEvm> => {
  const logger = createLogger({
    name: '@tevm/trevm',
    level: loggingLevel ?? 'warn',
  });
  
  logger.debug({
    allowUnlimitedContractSize,
    profiler,
    customPrecompiles: customPrecompiles?.map(c => c.address.toString()),
  });

  try {
    // Initialize the native trevm module with callbacks to the StateManager
    const nativeTrevm = await init_trevm({
      chainId: common.chainId(),
      allowUnlimitedContractSize: allowUnlimitedContractSize ?? false,
      stateCallbacks: {
        getAccount: async (address: string) => {
          try {
            const account = await stateManager.getAccount(address);
            if (!account) return null;
            
            return {
              nonce: account.nonce,
              balance: account.balance,
              codeHash: account.codeHash ? `0x${Buffer.from(account.codeHash).toString('hex')}` : undefined,
              code: account.code ? `0x${Buffer.from(account.code).toString('hex')}` : undefined,
            };
          } catch (error) {
            logger.error(`Error in getAccount for ${address}:`, error);
            return null;
          }
        },
        putAccount: async (address: string, account: any) => {
          try {
            await stateManager.putAccount(address, {
              nonce: account.nonce,
              balance: account.balance,
              codeHash: account.codeHash ? Buffer.from(account.codeHash.slice(2), 'hex') : undefined,
              code: account.code ? Buffer.from(account.code.slice(2), 'hex') : undefined,
            });
            return true;
          } catch (error) {
            logger.error(`Error in putAccount for ${address}:`, error);
            return false;
          }
        },
        getStorage: async (address: string, key: string) => {
          try {
            const value = await stateManager.getContractStorage(address, key);
            return value ? `0x${Buffer.from(value).toString('hex')}` : '0x0';
          } catch (error) {
            logger.error(`Error in getStorage for ${address} key ${key}:`, error);
            return '0x0';
          }
        },
        putStorage: async (address: string, key: string, value: string) => {
          try {
            await stateManager.putContractStorage(
              address,
              key,
              Buffer.from(value.slice(2), 'hex')
            );
            return true;
          } catch (error) {
            logger.error(`Error in putStorage for ${address} key ${key}:`, error);
            return false;
          }
        },
        getCode: async (address: string) => {
          try {
            const code = await stateManager.getContractCode(address);
            return code ? `0x${Buffer.from(code).toString('hex')}` : '0x';
          } catch (error) {
            logger.error(`Error in getCode for ${address}:`, error);
            return '0x';
          }
        },
        putCode: async (address: string, code: string) => {
          try {
            await stateManager.putContractCode(
              address,
              Buffer.from(code.slice(2), 'hex')
            );
            return true;
          } catch (error) {
            logger.error(`Error in putCode for ${address}:`, error);
            return false;
          }
        },
        getBlockHash: async (blockNumber: number) => {
          try {
            const block = await blockchain.getBlock(BigInt(blockNumber));
            return block ? `0x${Buffer.from(block.header.hash()).toString('hex')}` : null;
          } catch (error) {
            logger.error(`Error in getBlockHash for block ${blockNumber}:`, error);
            return null;
          }
        }
      }
    });

    // Return a JavaScript wrapper around the native trevm that matches the Evm interface
    return {
      async runCall(opts: TrevmRunCallOptions): Promise<TrevmRunCallResult> {
        try {
          return await nativeTrevm.run_call({
            caller: opts.caller,
            to: opts.to,
            data: opts.data,
            value: opts.value,
            gas_limit: opts.gasLimit,
            skip_balance: opts.skipBalance,
          });
        } catch (error) {
          logger.error('Error in runCall:', error);
          throw new InternalError(`Trevm execution error: ${error}`);
        }
      }
    };
  } catch (error) {
    logger.error('Error initializing trevm:', error);
    throw new InternalError(`Failed to initialize trevm: ${error}`);
  }
};