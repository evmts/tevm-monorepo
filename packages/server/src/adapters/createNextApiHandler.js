import { createHttpHandler } from '../createHttpHandler.js'

/**
 * Creates a Next.js API handler for a Tevm JSON-RPC server
 * @param {Pick<import('@tevm/memory-client').MemoryClient, 'send'>} client
 * @returns {import('next').NextApiHandler}
 * @example
 * ```typescript
 * import { createNextApiHandler } from 'tevm/server'
 * import { createMemoryClient } from 'tevm'
 *
 * const tevm = createMemoryClient()
 * export default createNextApiHandler({ request: tevm.request })
 * ```
 */
export const createNextApiHandler = (client) => {
	const handler = createHttpHandler(client)
	return async (req, res) => {
		handler(req, res)
	}
}
