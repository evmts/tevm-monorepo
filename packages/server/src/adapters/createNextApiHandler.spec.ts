import { createServer } from 'node:http'
import { createMemoryClient } from '@tevm/memory-client'
import type { NextApiHandler } from 'next'
import supertest from 'supertest'
import { beforeEach, describe, expect, it } from 'vitest'
import { createNextApiHandler } from './createNextApiHandler.js'

describe('createNextApiHandler', () => {
	let handler: NextApiHandler

	beforeEach(() => {
		handler = createNextApiHandler(createMemoryClient())
	})

	it('should handle valid JSON-RPC request', async () => {
		const server = createServer((req, res) => handler(req as any, res as any))

		const req = {
			jsonrpc: '2.0',
			method: 'eth_chainId',
			params: [],
			id: 1,
		} as const

		const res = await supertest(server).post('/').send(req).expect(200).expect('Content-Type', /json/)

		expect(res.body).toEqual({
			jsonrpc: '2.0',
			id: 1,
			method: 'eth_chainId',
			result: '0x384',
		})
	})

	it('should return 400 for invalid JSON', async () => {
		const server = createServer((req, res) => handler(req as any, res as any))

		const invalidJson = '{ "jsonrpc": "2.0", "method": "eth_chainId", "params": [ ] '

		const res = await supertest(server).post('/').send(invalidJson).expect(400).expect('Content-Type', /json/)

		expect(res.body.error).toBeDefined()
		expect(res.body.error.code).toBe(-32700)
		expect(res.body.error.message).toMatchSnapshot()
	})

	it('should return 400 for invalid JSON-RPC request', async () => {
		const server = createServer((req, res) => handler(req as any, res as any))

		const invalidRpcRequest = { jsonrpc: '2.0', method: 'invalid_method', params: 'invalid_params', id: 1 }

		const res = await supertest(server).post('/').send(invalidRpcRequest).expect(400).expect('Content-Type', /json/)

		expect(res.body.error).toBeDefined()
		expect(res.body.error.code).toBe(-32601)
		expect(res.body.error.message).toMatchSnapshot()
	})
})
