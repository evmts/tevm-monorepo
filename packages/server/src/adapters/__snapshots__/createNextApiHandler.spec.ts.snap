// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createNextApiHandler > should return 400 for invalid JSON 1`] = `
"Expected ',' or '}' after property value in JSON at position 59 (line 1 column 60)

Docs: https://tevm.sh/reference/tevm/errors/classes/invalidjsonerror/
Details: Expected ',' or '}' after property value in JSON at position 59 (line 1 column 60)
Version: 1.1.0.next-73"
`;

exports[`createNextApiHandler > should return 400 for invalid JSON-RPC request 1`] = `
"UnsupportedMethodError: Unknown method invalid_method

Docs: https://tevm.sh/reference/tevm/errors/classes/methodnotfounderror/
Version: 1.1.0.next-73"
`;
